Reenable UDP stunner and add documentation

sehraf opened the pull request:
Retroshare implements the STUN protocol (old RFC 3489) to determine the NAT type it is running behind.
While the actual determination logic resides in the class `pqiNetStateBox`, the UDP stunner is responsible for running the actual protocol.

The UDP stunner was disabled a while ago because it seemed only be used to determine the external IP address (using STUN). While this is true as well it's only a byproduct.
As @drbob pointed out it should be re-enabled to allow RS to determine the NAT type again - and based on that knowledge - build more UDP connections (there are multiple types and currently only the one (*direct*) is working).

The patch further adds some (minor) documentation to the UDP stunner, makes the implementation more RFC conform as well as document the RFCs responsible and allows (optionally) to utilize the it for external IP determination. (While this is generally problematical it might be useful for corner cases)

sehraf committed:
enable DHT stunner

sehraf committed:
first pass on udpstunner.cc

sehraf committed:
add some documentation to updstunner.h

sehraf committed:
udpstunner: don't open code header fields, reference and follow (old) RFC

sehraf committed:
* unify CONFIG option for DHT stunner
* implement optional (turned off by default) usage of dht stunner to determin external ip (this can be usefull for some corner case setups)

sehraf committed:
add more documentation to updstunner.h

sehraf committed:
review fixup

G10h4ck commented on libretroshare/src/tcponudp/udpstunner.cc:
```diff
@@ -47,6 +47,12 @@ const uint32_t TOU_STUN_MAX_RECV_RATE = 25; /* every 25 seconds */
+ * based on RFC 3489
+ */
+const uint16_t STUN_BINDING_REQUEST  = 0x0001;
+const uint16_t STUN_BINDING_RESPONSE = 0x0101;
```
Wouldn't it be more appropriate to have those as static constxepr memebers of UdpStunner ?

sehraf replied:
Sure, will do.

G10h4ck commented on libretroshare/src/tcponudp/udpstunner.cc:
```diff
@@ -517,33 +523,41 @@ bool UdpStun_generate_stun_pkt(void *stun_pkt, int *len)
+	 *	The transaction ID is used to correlate requests and responses.
+	 *
+	 * RFC 5389 introduces a mmgic cokie at the location where preciously the transaction ID was located:
+	 *	In RFC 3489 [RFC3489], this field was part of
```
why is RFC3489 repeated twice?

sehraf replied:
It's a plain copy+paste from the RFC.
Can remove the reference in square brackets.

G10h4ck commented on libretroshare/src/tcponudp/udpstunner.cc:
```diff
@@ -517,33 +523,41 @@ bool UdpStun_generate_stun_pkt(void *stun_pkt, int *len)
+	((uint32_t *) stun_pkt)[1] = htonl(RsRandom::random_u32());
+	((uint32_t *) stun_pkt)[2] = htonl(RsRandom::random_u32());
+	((uint32_t *) stun_pkt)[3] = htonl(RsRandom::random_u32());
+	((uint32_t *) stun_pkt)[4] = htonl(RsRandom::random_u32());
```
This 4 lines would e probably easier substituted by a call of `RsRandom::random_bytes`

sehraf replied:
Can do

G10h4ck commented on libretroshare/src/tcponudp/udpstunner.h:
```diff
@@ -53,10 +53,15 @@ class TouStunPeer
+	/// id for identification
 	std::string id;
+	/// remote address
 	struct sockaddr_in remote, eaddr;
```
Please document both members

sehraf replied:
Ah, oversaw that! Thanks for the review!

sehraf commented:
I've rebased and squashed the fixups.

G10h4ck commented:
Now it looks ok to me, I'd wait a few days for @csoler and @drbob to say something, and then merge it

csoler commented:
What is the priority of the external IP information using STUN w.r.t. other---more reliable---methods? Normally the expected behavior should be that external IP given by friends (that is 100% reliable) should erase previous estimates given by STUN. Is that the case already?

sehraf commented:
When you enable this mechanic (it's off by default) it counts as one (like all other sources, I think)
This happens as soon as **two** dht peers have reported the same external address.

Edit:
If desired I can change it to only use it as a fallback.

sehraf commented:
For the record:
* you must use `CONFIG+=use_dht_stunner_ext_ip` to use the dht stunner as a source of external IP
* dht stunner will only return an external IP when it was reported by two peers
* the switch turns on the old behavior 
* it counts as one like all other sources do, see https://github.com/RetroShare/RetroShare/blob/6f9d875fc69fab52b20e5f886fc614949d937250/libretroshare/src/pqi/p3netmgr.cc#L666

csoler commented:
> When you enable this mechanic (it's off by default) it counts as one (like all other sources, I think)
> This happens as soon as **two** dht peers have reported the same external address.
> 
> Edit:
> If desired I can change it to only use it as a fallback.

Well, given that DHT is poisoned by peers who impersnnate your friends with a wrong IP, I am a bit skeptical about the effect on the DHT stunner and the reliability of the IP it will give. So it's important that other more reliable sources (such as discovery) wipe out the result of DHT stunner when you use it.

sehraf committed:
only use external up from DHT stunner as fallback when no other source provides a valid IP

sehraf commented:
I've moved the block that gets the external IP from DHT stunner to the end and added a `if (address_votes.empty()) ` to only use it as fallback when all others haven't provided an IP address.
